runtime/sema/type_tags.go

```diff
@@ -404,10 +423,16 @@ func findCommonSuperType(joinedTypeTag TypeTag, types ...Type) Type {
 	// finding super type for the rest of the types is sufficient.
 	joinedTypeTag = joinedTypeTag.And(notNeverType)
 
-	if joinedTypeTag.upperMask != 0 {
-		superType = findSuperTypeFromUpperMask(joinedTypeTag, types)
-	} else {
+	if joinedTypeTag == NoTypeTag {
+		return InvalidType
+	}
+
+	// If both masks are on, then the types are heterogeneous.
+	// So skip the optimization and find the supertype in the hard way.
+	if joinedTypeTag.lowerMask != 0 && joinedTypeTag.upperMask == 0 {
 		superType = findSuperTypeFromLowerMask(joinedTypeTag, types)
+	} else if joinedTypeTag.lowerMask == 0 && joinedTypeTag.upperMask != 0 {
+		superType = findSuperTypeFromUpperMask(joinedTypeTag, types)
```

Comment:
The actual change/fix is here.
Earlier it was assumed that checking the existence of 'upper-mask' is sufficient, because the upper mask only had three types, and for those three types `'findSuperTypeFromUpperMask' also handled heterogeneous types.
However, when adding more types to upper-masks, this duplicates the work as they all needed to handle heterogeneous types for newly added types as well.

Instead, if both upper and lower masks are present, fall back to the default case here, so `findSuperTypeFromUpperMask` don't have to worry about heterogeneous types. Heterogeneous types are handled outside of this 'if' condition already.